### PR TITLE
Validation summary displays generic error messages for ignored exceptions

### PR DESCRIPTION
**Pull request by ajaybhargavb**
#3043 

@dougbu / @NTaylorMullen

### CODE REVIEW

File: test/Microsoft.AspNet.Mvc.ViewFeatures.Test/Rendering/HtmlHelperValidationSummaryTest.cs [hunk 1]
```diff
@@ -342,21 +347,21 @@ private void AddMultipleErrors(ModelStateDictionary modelState)
             modelState.AddModelError("Property3.OrderedProperty2", "This is an error for Property3.OrderedProperty2.");
 
             modelState.AddModelError("Property3", "This is an error for Property3.");
-            modelState.AddModelError("Property3", new InvalidCastException("Exception will be ignored."));
+            modelState.AddModelError("Property3", new InvalidCastException("Produces generic error message"));
```
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.ViewFeatures.Test/Rendering/HtmlHelperValidationSummaryTest.cs]: Why change these ignored messages?

Comment (ajaybhargavb) [inline test/Microsoft.AspNet.Mvc.ViewFeatures.Test/Rendering/HtmlHelperValidationSummaryTest.cs]: Because they are not actually ignored but are displayed as generic errors.

File: src/Microsoft.AspNet.Mvc.ViewFeatures/ViewFeatures/DefaultHtmlGenerator.cs [hunk 2]
```diff
@@ -802,7 +802,7 @@ public virtual IHtmlContent GenerateAntiforgery(ViewContext viewContext)
             {
                 foreach (var modelError in modelState.Errors)
                 {
-                    var errorText = ValidationHelpers.GetUserErrorMessageOrDefault(modelError, modelState: null);
+                    var errorText = ValidationHelpers.GetUserErrorMessageOrDefault(modelError, modelState);
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.ViewFeatures/ViewFeatures/DefaultHtmlGenerator.cs]: This is a change from validation summaries provided in MVC 5 and earlier. It's been consistent for validation messages (at field level) to include default "is invalid" message but to hide such messages in validation summaries. What I expected to fix #3043 was another special case in `TryAddModelError(... Exception ...)` for the  `Exception` type thrown in that scenario (`OverflowException`?), similar to `FormatException` handling but with a different message e.g. `Value '{0}' is out of range for '{1}'.`. @Eilon your thoughts?

If we decide to use current approach, should remove the `modelState == null` special case in `GetUserErrorMessageOrDefault()`. Also need an integration test demonstrating new behaviour for exact #3043 scenario (attempt to bind a negative value to a `uint` property).

File: test/Microsoft.AspNet.Mvc.ViewFeatures.Test/Rendering/HtmlHelperValidationSummaryTest.cs (hunk 1 above)
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.ViewFeatures.Test/Rendering/HtmlHelperValidationSummaryTest.cs]: You're quibbling but :ok:

File: src/Microsoft.AspNet.Mvc.ViewFeatures/ViewFeatures/DefaultHtmlGenerator.cs (hunk 2 above)
Comment (ajaybhargavb) [inline src/Microsoft.AspNet.Mvc.ViewFeatures/ViewFeatures/DefaultHtmlGenerator.cs]: Checked MVC 5 behavior. It does not ignore the errors in validation summary.

image


Looked at the code and there is a special casing in `DefaultModelBinder` for Format and Overflow exceptions which adds the invalid messages. http://aspnetwebstack.codeplex.com/SourceControl/latest#src/System.Web.Mvc/DefaultModelBinder.cs
```C#
// We only consider "known" type of exception and do not make too aggressive changes here
if (exception is FormatException || exception is OverflowException)
{
    string displayName = propertyMetadata.GetDisplayName();
    string errorMessageTemplate = GetValueInvalidResource(controllerContext);
    string errorMessage = String.Format(CultureInfo.CurrentCulture, errorMessageTemplate, modelState.Value.AttemptedValue, displayName);
    modelState.Errors.Remove(error);
    modelState.Errors.Add(errorMessage);
    break;
}
```

Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.ViewFeatures/ViewFeatures/DefaultHtmlGenerator.cs]: Let's start with fixing that issue.

We can research why the validation summary has historically been shorter than the sum of all validation messages and tackle that separately. Worth filing a separate Investigate item for that.

Comment (NTaylorMullen) [inline src/Microsoft.AspNet.Mvc.ViewFeatures/ViewFeatures/DefaultHtmlGenerator.cs]: :+1:

**Comment by ajaybhargavb** [comment]
Updated.

**Comment by NTaylorMullen** [comment]
Since the new change addresses [this](https://github.com/aspnet/Mvc/pull/3214#discussion_r40627250). :shipit:. 

Should track investigation of the second part of that comment to determine what we should or should not hide in MVC6.

File: test/Microsoft.AspNet.Mvc.IntegrationTests/ValidationIntegrationTests.cs [hunk 3]
```diff
@@ -1086,5 +1162,13 @@ private static void AssertRequiredError(string key, ModelError error)
                 string.Format("The {0} field is required.", key)), error.ErrorMessage);
             Assert.Null(error.Exception);
         }
+
+        private class User
```
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.IntegrationTests/ValidationIntegrationTests.cs]: Everywhere else in this test class, the target types are defined just before their first use. Not good practise in product code but it makes this long file easier to use. So move this above `Validation_FormatException_ShowsInvalidValueMessage_OnSimpleTypeProperty()`.

File: test/Microsoft.AspNet.Mvc.IntegrationTests/ValidationIntegrationTests.cs [hunk 4]
```diff
@@ -992,6 +992,82 @@ private class Order10
...
+            Assert.Equal(0, model.Id);
+
+            Assert.Equal(1, modelState.Count);
+            Assert.Equal(1, modelState.ErrorCount);
+            Assert.False(modelState.IsValid);
+
+            var entry = Assert.Single(modelState, e => e.Key == "Id").Value;
+            Assert.Equal("bill", entry.AttemptedValue);
+            Assert.Equal("bill", entry.RawValue);
+            Assert.Single(entry.Errors);
```
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.IntegrationTests/ValidationIntegrationTests.cs]: Better to use `var error =` here than `entry.Errors[0]` on the next line.

File: test/Microsoft.AspNet.Mvc.IntegrationTests/ValidationIntegrationTests.cs [hunk 5]
```diff
@@ -992,6 +992,82 @@ private class Order10
...
+            Assert.True(modelBindingResult.IsModelSet);
+
+            var model = Assert.IsType<User>(modelBindingResult.Model);
+            Assert.Equal(0, model.Id);
+
+            Assert.Equal(1, modelState.Count);
+            Assert.Equal(1, modelState.ErrorCount);
+            Assert.False(modelState.IsValid);
+
+            var entry = Assert.Single(modelState, e => e.Key == "Id").Value;
```
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.IntegrationTests/ValidationIntegrationTests.cs]: Checks here are redundant. Don't need `Assert.Equal(1, modelState.Count)`:
``` c#
var state = Assert.Single(modelState);
Assert.Equal("Id", state.Key);
var entry = state.Value;
```

**Comment by dougbu** [comment]
:shipit:
